js/node_shared/unix.js

```diff
@@ -0,0 +1,348 @@
+/* What is this file for?
+   We would like to transpile our OCaml code to JS, to make it runnable in
+   the browser and on Windows systems. The OCaml standard library, however,
+   has external C dependencies, some of which are unimplemented by the JSOO
+   transpiler.
+   In particular, the Unix module consists largely of external C calls which
+   are not defined symbols in the produced JS.
+   To make sure we can still call those C functions (which happens in things
+   such as Bos, Time_limit, etc), we reimplement those fundamental Unix calls
+   in Javascript.
+   Note that this doesn't need to be a perfect reimplementation, but we can
+   get by with "good enough" behavior.
+ */
+
+const { spawnSync } = require("node:child_process");
+const os = require("os");
+const fs = require("fs");
+const process = require("process");
+
+globalThis.spawnSync = spawnSync;
+globalThis.fs = fs;
+globalThis.os = os;
+globalThis.process = process;
+
+// mapping pid to its own process
+globalThis.pidToProcessTable = new Map();
+// mapping pid of parent to its children
+globalThis.pidToChildrenTable = new Map();
+// There will be collisions if we ever allocate 20000 or more file descriptors
+// "naturally", that is, through `caml_sys_open`.
+// "Fake", because these file descriptors we produce are not real file descriptors,
+// but just indices corresponding to `unix_spawn` calls.
+// According to Austin: should be Ok, because node uses libuv
+// https://www.geeksforgeeks.org/libuv-in-node-js/
```

Comment:
Can you help me understand what you mean by collisions here? The OS selects the lowest unused number when allocating an fd, so I can see how collisions can happen if you're storing the fd numbers in a map, but I don't really understand what the problem is, and how limiting the lifetime number of fds allocated will help here.

Reply:
Resolved in call, and added a clarifying comment